File: top_100_liked/python3/search_in_rotated_sorted_array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        if(n <= 1):
            return -1 if (n == 0 or target != nums[0]) else 0
        if(nums[0] < nums[-1]):
            k = 0
        else:
            ### search for k
            l, r = 0, n - 1 # since there is always a num satisfying the condition
            while(l < r):
                mid = (l+r)//2
                if(nums[mid] <= nums[-1]):
                    r = mid
                else:
                    l = mid + 1
            k = n - l
        ### normal binary search with mapping
        l, r = 0, n # might not find val satisfying val >= target (l would be n)
        while(l < r):
            mid = (l+r)//2
            val = nums[(mid - k + n) % n]
            if(val >= target):
                r = mid
            else:
                l = mid + 1
        if(l < n and nums[(l - k + n) % n] == target):
            return (l - k + n) % n
        return -1

File: top_100_liked/cpp/search_in_rotated_sorted_array.cpp
class Solution {
public:
    // find k version
    int v1search(vector<int>& nums, int target) {
        // first binary search to find k
        int k, n = nums.size();
        int mid, l = 0, r = n;
        if(nums[n-1] >= nums[0])
            k = 0;
        else{
            while(l < r){
                mid = (r-l)/2 + l;
                if(mid-1 >= 0 && nums[mid] < nums[mid-1]){
                    k = mid;
                    break;
                }
                else{
                    if(nums[mid] < nums[0])
                        r = mid;
                    else
                        l = mid + 1;
                }
            }
        }
        // normal binary search with index-mapping
        l = 0;
        r = n;
        while(l < r){
            mid = (r-l)/2 + l;
            if(nums[(mid + k) % n] == target)
                return (mid + k) % n;
            else if(nums[(mid + k) % n] > target)
                r = mid;
            else
                l = mid + 1;
        }
        return -1;
    }
    // direct search
    int search(vector<int>& nums, int target){
        
    }
};
